metaseq/models/transformer.py

```diff
@@ -561,10 +562,35 @@ def forward_embedding(
         incremental_state: Optional[Dict[str, Dict[str, Optional[Tensor]]]] = None,
     ):
         # embed tokens and positions
-        positions = None
+        if self.self_attn_doc_sep != -1:
+            # create own positions when self_attn_doc_sep is set
+            mask = tokens.ne(self.padding_idx).int()
+            mask_with_reset = tokens.ne(self.padding_idx).int()
+            mask_with_reset[:, :] = 1
+            doc_id_indices = (tokens == self.self_attn_doc_sep).nonzero().tolist()
+            for batch_idx in range(tokens.size(0)):
+                batch_doc_indices = [
```

Comment:
Nit: add a comment for what this logic is doing (separating intention vs the implementation) to make it easier to parse what nested for loops are trying to loop over.